adr/ADR-15.md

```diff
@@ -0,0 +1,233 @@
+
+Some information allowing proper routing in context of multiple accounts and preventing an ACK for an account to ACK a message for the same stream and consumer names in another, were missing.
+
+Until now, ACK reply subject contained 9 tokens, with this layout:
+```
+$JS.ACK.<stream name>.<consumer name>.<num delivered>.<stream sequence>.<consumer sequence>.<timestamp>.<num pending>
+```
+
+This is going to change with newer servers where the number of tokens will be 12, with the tokens being:
+```
+$JS.ACK.<domain>.<account hash>.<stream name>.<consumer name>.<num delivered>.<stream sequence>.<consumer sequence>.<timestamp>.<num pending>.<a token with a random value>
+```
+
+When there is no domain, the server will still set the token to a special value of `_` (the server will make sure that users can't pick this as a domain name). Having the domain always present simplifies the library code which does not have to bother with a variable number of tokens somewhere close to the beginning of the subject, and possibly doing some shifting to find out the location of the fields it cares about.
+
+Why not append those new tokens at the end? This is to simplify the export/import subject, ie `$JS.ACK.<domain>.<account>.>`. Otherwise, you would need to possibly have something like `$JS.ACK.*.*.*.*.*.*.*.<domain>.<account>.>`
+
+When unpacking an ACK subject, the library should verify that it starts with `$JS.ACK.` but then check the overall number of tokens, and if 9 assume that it is the "old" ACK messages without domain nor account hash.
+
+If it has at *least* 12 tokens, we know that 3rd token is domain, and is reported as a new field `Domain` in the `MsgMetadata` object. The library can replace `_` with empty string when returning a `MsgMetada` object, or document the meaning of the domain named `_`.
+
+The account hash is not used by the client at this time, only used for routing, same for the last token.
+
+It is recommended that library no longer report an error if there are more than 12 tokens, so that new tokens may be added, especially if their purpose is solely for routing and have no impact on the library itself.
+
+#### Flow control
+
+When a JetStream consumer has flow control enabled (not applicable to pull consumers), the library may receive a control message that indicates that this is a flow control and should respond to the provided subject when the current pending message has been processed.
+
+That is, suppose messages are received on a connection and pushed to a JetStream subscription, and there are currently 1000 pending messages. When the connection received a JetStream flow control status message for a given subscription, it should mark that the subscription should send an empty message to the given Flow Control subject (which is the reply subject of the incoming control message) when the library has dispatched that 1000's message in the subscription's pending list.
```

Comment:
Just to make sure this is not happening: The format or the number of tokens of the Fc message is not to be inspected.

Reply:
Added a note.